### PR TITLE
🚨 Add training compatibility for Musicgen-like models

### PR DESCRIPTION
**Pull request by ylacombe**
This PR aims to add training compatibility for Musicgen and Musicgen Melody.

The main difference with classic cross-entropy is that there a `num_codebooks` labels to predict per timestamp instead of a single token per timestamp. This materializes in the loss which consists in the mean of cross-entropy per codebook.

A few additional insights:
- The models don't have an EOS token id, so the models generate for `max_length`.
- The model actually predict codebooks in a delayed pattern.
            - The first codebook channel is predicted without delay, but the further you go, the more delay there is (2nd codebook -> delayed by 1, 3rd codebook -> delayed by 2, etc.)
- Training scripts will be shared as well

cc @sanchit-gandhi and @amyeroberts

### CODE REVIEW

**Comment by arjunsinghrathore** [comment]
Hi! Is it possible to finetune the musicgen model currently? If yes then is there something I should keep in mind. Would be really helpful if you could share your opinions. Thanks!

**Comment by LiuZH-19** [comment]
Wonderful work! I'm currently attempting to fine-tune the Musicgen model using these codes, but I haven't succeeded yet. Is the model ready for fine-tuning, and are there specific aspects I should be aware of? Any training tips or guidance you could provide would be greatly appreciated!

Thank you so much!

**Comment by ylacombe** [comment]
Hey @arjunsinghrathore and @LiuZH-19, I'll likely release some fine-tuning code next week or the week after! 
May I ask what type of data do you have, out of curiosity ?
Thanks!

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 1]
```diff
@@ -1340,15 +1343,22 @@ def forward(
         return_dict: Optional[bool] = None,
     ) -> Union[Tuple, CausalLMOutputWithCrossAttentions]:
         r"""
-        labels (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
+        labels (`torch.LongTensor` of shape `(batch_size, sequence_length, num_codebooks)`, *optional*):
             Labels for language modeling. Note that the labels **are shifted** inside the model, i.e. you can set
             `labels = input_ids` Indices are selected in `[-100, 0, ..., config.vocab_size]` All labels set to `-100`
```
Comment (sanchit-gandhi) [inline src/transformers/models/musicgen/modeling_musicgen.py:1348]: _c.f._ this docstring that explains how the `-100` padding token should be set

File: tests/models/musicgen/test_modeling_musicgen.py [hunk 2]
```diff
@@ -189,6 +198,47 @@ def setUp(self):
...
+        config.use_cache = False
+        config.return_dict = True
+        model = MusicgenForCausalLM(config)
+
+        model.to(torch_device)
+        model.gradient_checkpointing_enable(gradient_checkpointing_kwargs=gradient_checkpointing_kwargs)
+        model.train()
+
+        # Contrarily to the initial method, we don't unfreeze freezed parameters.
+        # Otherwise, it'll mess with the freezed sinusoidal embeddings
```
Comment (sanchit-gandhi) [inline tests/models/musicgen/test_modeling_musicgen.py:229]: It'll change the weights of the pre-trained embeddings for sure, but the code should still run right? Or is there an issue where the code won't run if we train the embeddings? Unless this is the case, I would just use the base `check_training_gradient_checkpointing` method for simplicity

File: src/transformers/models/musicgen_melody/modeling_musicgen_melody.py [hunk 3]
```diff
@@ -2428,6 +2451,16 @@ def _maybe_initialize_input_ids_for_generation(
                 break
         return torch.ones((batch_size, 1), dtype=torch.long, device=self.device) * bos_token_id
 
+    def freeze_encoders(self, freeze_text_encoder=True):
```
Comment (sanchit-gandhi) [inline src/transformers/models/musicgen_melody/modeling_musicgen_melody.py:2454]: You can also do a simple test to confirm this method works as expected: https://github.com/huggingface/transformers/blob/440bd3c3c06004115752c57530c7150b38f2a0a7/tests/models/whisper/test_modeling_whisper.py#L435

**Review by sanchit-gandhi — APPROVED** [review]
Fast tests look quite rigorous, but we can probably use the base one unless there's a specific embedding issue! Otherwise LGTM

File: tests/models/musicgen/test_modeling_musicgen.py (hunk 2 above)
Comment (ylacombe) [reply tests/models/musicgen/test_modeling_musicgen.py:229]: The comments are not really clear, I'll make it clearer: the ConditionalGeneration models have the audio encoder which is never used outside of the `.generate`, if we don't freeze it, we'll have some issues because trainable weights won't have seen any gradients.

**Comment by ylacombe** [comment]
Hey @amyeroberts, gentle ping to ask for a review! Many thanks for your help!

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 4]
```diff
@@ -1340,15 +1345,18 @@ def forward(
...
             are ignored (masked), the loss is only computed for labels in `[0, ..., config.vocab_size]`
-                Returns:
+        Returns:
         """
 
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
+        if (labels is not None) and (input_ids is None and inputs_embeds is None):
+            input_ids = shift_tokens_right(labels, self.config.pad_token_id, self.config.bos_token_id)
```
Comment (amyeroberts) [inline src/transformers/models/musicgen/modeling_musicgen.py:1358]: It's a bit funny to do this on this input_ids - normally it's just on the decoder_input_ids.

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 5]
```diff
@@ -1370,7 +1380,28 @@ def forward(
...
+
+            # per codebook cross-entropy
+            for codebook in range(self.config.num_codebooks):
+                codebook_logits = logits[:, codebook].contiguous().view(-1, logits.shape[-1])
+                codebook_mask = mask[..., codebook].contiguous().view(-1)
+                codebook_labels = labels[..., codebook].contiguous().view(-1)
+
+                loss += loss_fct(codebook_logits[codebook_mask], codebook_labels[codebook_mask])
+
+            loss = loss / self.config.num_codebooks
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/musicgen/modeling_musicgen.py:1398]: Let's add this link in a comment above for any unsuspecting future code reader to have context

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 6]
```diff
@@ -2270,23 +2295,15 @@ def forward(
...
         if not return_dict:
-            if loss is not None:
-                return (loss,) + decoder_outputs + encoder_outputs
-            else:
-                return decoder_outputs + encoder_outputs
+            return decoder_outputs + encoder_outputs
 
         return Seq2SeqLMOutput(
-            loss=loss,
+            loss=decoder_outputs.loss,
```
Comment (amyeroberts) [inline src/transformers/models/musicgen/modeling_musicgen.py:2306]: hmmm, the problem with this is it's not backwards compatible - users are now going to get different values of loss than before

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 7]
```diff
@@ -2533,6 +2550,16 @@ def resize_token_embeddings(self, *args, **kwargs):
...
 
+    def freeze_encoders(self, freeze_text_encoder=True):
+        if freeze_text_encoder:
+            for param in self.text_encoder.parameters():
+                param.requires_grad = False
+            self.text_encoder._requires_grad = False
+
+        for param in self.audio_encoder.parameters():
+            param.requires_grad = False
+        self.audio_encoder._requires_grad = False
```
Comment (amyeroberts) [inline src/transformers/models/musicgen/modeling_musicgen.py:2561]: I'm not a fan of this structure. In general, we don't add `freeze` methods to our models add leave that to the user to handle - although I see audio models appear to be the exception! 

It's tidier to split this up to `freeze_text_encoder` and `freeze_audio_encoder` and then just call them separately or add an additional `freeze_audio_encoder` argument

File: src/transformers/models/musicgen_melody/modeling_musicgen_melody.py [hunk 8]
```diff
@@ -2428,6 +2442,16 @@ def _maybe_initialize_input_ids_for_generation(
...
```
Comment (amyeroberts) [inline src/transformers/models/musicgen_melody/modeling_musicgen_melody.py:2445]: Same comment here

File: tests/models/musicgen/test_modeling_musicgen.py [hunk 9]
```diff
@@ -189,6 +198,47 @@ def setUp(self):
...
+        inputs_dict = super()._prepare_for_class(inputs_dict, model_class, return_labels=return_labels)
+
+        if return_labels:
+            inputs_dict["labels"] = torch.zeros(
+                (self.model_tester.batch_size, self.model_tester.seq_length, self.model_tester.num_codebooks),
+                dtype=torch.long,
+                device=torch_device,
+            )
+        else:
+            inputs_dict.pop("labels", None)
```
Comment (amyeroberts) [inline tests/models/musicgen/test_modeling_musicgen.py:212]: Why would you need to pop if `return_labels` is passed in the `super()._prepare_for_class` call?

File: tests/models/musicgen/test_modeling_musicgen.py [hunk 10]
```diff
@@ -1032,6 +1084,49 @@ class MusicgenTest(ModelTesterMixin, GenerationTesterMixin, PipelineTesterMixin,
...
```
Comment (amyeroberts) [inline tests/models/musicgen/test_modeling_musicgen.py:1098]: Same q here

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 11]
```diff
@@ -2533,6 +2550,16 @@ def resize_token_embeddings(self, *args, **kwargs):
             " model.decoder.resize_token_embeddings(...))"
         )
 
+    def freeze_encoders(self, freeze_text_encoder=True):
```
Comment (amyeroberts) [inline src/transformers/models/musicgen/modeling_musicgen.py:2553]: Should have a docstring as its a public method

File: tests/models/musicgen/test_modeling_musicgen.py [hunk 12]
```diff
@@ -2189,6 +2288,26 @@ def test_eager_matches_sdpa_generate(self):
 
                 self.assertTrue(torch.allclose(res_eager, res_sdpa))
 
+    def test_requires_grad_with_frozen_encoders(self):
```
Comment (amyeroberts) [inline tests/models/musicgen/test_modeling_musicgen.py:2291]: ❤️

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this capability! 

Mostly small comments. Only concern is backwards compatibility wrt the loss for musicgen

File: src/transformers/models/musicgen/modeling_musicgen.py (hunk 4 above)
Comment (ylacombe) [reply src/transformers/models/musicgen/modeling_musicgen.py:1358]: Well, it's a bit confusing but `input_ids` in MusicgenForCausalLM actually corresponds to the audio input_ids (i.e the decoder input ids)!

File: src/transformers/models/musicgen/modeling_musicgen.py (hunk 6 above)
Comment (ylacombe) [reply src/transformers/models/musicgen/modeling_musicgen.py:2306]: I haven't seen any users using this or mentioning this tbh (besides it's totally wrong!).
How should we best handle this ? maybe adding a breaking flag in the PR name ?

File: src/transformers/models/musicgen/modeling_musicgen.py (hunk 7 above)
Comment (ylacombe) [reply src/transformers/models/musicgen/modeling_musicgen.py]: I've split these up !

**Comment by ylacombe** [comment]
Many thanks for the review @amyeroberts,  I've changed the code according to your comments!
The only left to address is the loss computation being breaking changes. let me know what you think of this.
Note that I don't believe a lot of users actually used the loss computation as it was.

File: src/transformers/models/musicgen/modeling_musicgen.py (hunk 4 above)
Comment (amyeroberts) [reply src/transformers/models/musicgen/modeling_musicgen.py:1358]: Ah, ok! Makes sense :)

File: src/transformers/models/musicgen/modeling_musicgen.py (hunk 6 above)
Comment (amyeroberts) [reply src/transformers/models/musicgen/modeling_musicgen.py:2307]: If it's completely wrong then I think it's OK to break. We should just add a 🚨  prefix to the PR title so it can be easily found when preparing the release notes

**Review by amyeroberts — APPROVED** [review]
Awesome work - thanks for adding this feature!
